ndb/src/google/cloud/ndb/model.py

```diff
@@ -1542,11 +1543,123 @@ def _get_for_dict(self, entity):
         return self._get_value(entity)
 
 
+def _validate_key(value, entity=None):
+    """Validate a key.
+
+    Args:
+        value (.Key): The key to be validated.
+        entity (Optional[Model]): The entity that the key is being validated
+            for.
+
+    Returns:
+        .Key: The passed in ``value``.
+
+    Raises:
+        .BadValueError: If ``value`` is not a :class:`.Key`.
+        KindError: If ``entity`` is specified, but the kind of the entity
+            doesn't match the kind of ``value``.
+    """
+    if not isinstance(value, Key):
+        raise exceptions.BadValueError("Expected Key, got {!r}".format(value))
+
+    if entity and type(entity) not in (Model, Expando):
```

Comment:
@chrisrossi I switched to `type(entity)` here. Should we clean up all usage of `.__class__` in this package?

Reply:
Yeah, probably.